modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/disruptor/inbound/InboundMessageRejectEvent.java

```diff
@@ -0,0 +1,115 @@
+import org.wso2.andes.kernel.AndesContext;
+import org.wso2.andes.kernel.AndesException;
+import org.wso2.andes.kernel.DeliverableAndesMetadata;
+import org.wso2.andes.kernel.subscription.AndesSubscription;
+import org.wso2.andes.tools.utils.MessageTracer;
+
+import java.util.UUID;
+
+/**
+ * Disruptor event handling message reject
+ */
+public class InboundMessageRejectEvent implements AndesInboundStateEvent {
+
+
+    private static Log log = LogFactory.getLog(InboundMessageRejectEvent.class);
+    /**
+     * Channel event type handle by the event object
+     */
+    private EventType eventType;
+
+    /**
+     * Id of the message rejected by client
+     */
+    private long messageId;
+
+    /**
+     * Id of the channel reject is received
+     */
+    private UUID channelId;
+
+    /**
+     * Whether to re-queue the message back to the subscriber
+     */
+    private boolean reQueue;
+
+    /**
+     * This is to handle the messages beyond the actual rollback event, which are rejected from the client side
+     * message buffer. True if the message was rejected after the last rollback event.
+     */
+    private boolean isMessageBeyondLastRollback;
+
+    /**
+     * Supported state events
+     */
+    public enum EventType {
+        /**
+         * Specific client channel close event
+         */
+        MESSAGE_REJECT_EVENT,
+    }
+
+    /**
+     * Create an instance of message reject event
+     *
+     * @param messageId Id of the message rejected
+     * @param channelId Id of the channel message reject received
+     * @param reQueue   whether to re queue message back to subscriber
+     */
+    public InboundMessageRejectEvent(long messageId, UUID channelId, boolean reQueue) {
+        this.messageId = messageId;
+        this.channelId = channelId;
+        this.reQueue = reQueue;
+    }
+
+    /**
+     * Prepare to reject message. This is done prior to inset the
+     * event to disruptor
+     *
+     * @param isMessageBeyondLastRollback True if the message was rejected after the last rollback event.
+     */
+    public void prepareToRejectMessage(boolean isMessageBeyondLastRollback) {
+        this.isMessageBeyondLastRollback = isMessageBeyondLastRollback;
+        this.eventType = EventType.MESSAGE_REJECT_EVENT;
+    }
+
+    @Override
+    public void updateState() throws AndesException {
+        AndesSubscription subscription = AndesContext.getInstance().
+                getAndesSubscriptionManager().getSubscriptionByProtocolChannel(channelId);
+        if (subscription != null) {
+            DeliverableAndesMetadata rejectedMessage = subscription.onMessageReject(messageId, reQueue);
+            rejectedMessage.setIsBeyondLastRollbackedMessage(isMessageBeyondLastRollback);
+            //Tracing message activity
+            MessageTracer.trace(rejectedMessage, MessageTracer.MESSAGE_REJECTED);
+        } else {
+            log.warn("Cannot handle reject. Subscription not found for channel "
```

Comment:
But subscription might have closed in a sudden. In such cases this can happen (by design).